flink-runtime/src/main/java/org/apache/flink/runtime/metrics/dump/MetricQueryService.java

```diff
@@ -109,7 +122,17 @@ public void onReceive(Object message) {
 				}
 			} else if (message instanceof CreateDump) {
 				MetricDumpSerialization.MetricSerializationResult dump = serializer.serialize(counters, gauges, histograms, meters);
-				getSender().tell(dump, getSelf());
+
+				int realMsgSize = dump.serializedMetrics.length;
+
+				if (realMsgSize > maximumFramesize) {
+					String overSizeErrorMsg = "The metric dump message size : " + realMsgSize
+						+ " exceeds the maximum akka framesize : " + maximumFramesize + ".";
+					LOG.error(overSizeErrorMsg);
+					getSender().tell(new Status.Failure(new IOException(overSizeErrorMsg)), getSelf());
```

Comment:
While better than crashing a more reasonable option might be to create a new dump with reduced number of metrics.

Reply:
or even pass in maxSize to the serialization logic and only serialize up to that point.

Reply:
If we pass incomplete data, it will cause the receiver to parse failure, which seems to cause users trouble?

Reply:
well then don't pass incomplete data. metrics are serialized one at a time, so you can just remember during serialization the last metrics that fit into the buffer, and once you go over the limit go back to that.
You can also use other interesting heuristics like dropping histograms first (since they are significantly larger).

Reply:
alright, this isn't so easy to implement. I would suggest as a simple counter-measure to retry the serialization without histograms, and if that fails as well to fail with an exception. Making this more robust can be handled in a follow-up.

Reply:
OK, I will try to implement it in a simple way.

Reply:
I want to confirm with you, here the serialize operation is called in the `MetricQueryService`, but if you want to unregister a metric, it will make the `MetricQueryService` depend on the `MetricRegistry` (this dependency is reasonable?), and the parameters required by the unregister method, they have not been saved in `MetricQueryService` classes. I am not sure if we want to introduce this complexity? Or still returning to the original idea to serialize only part of the metrics?